File: daedalos-tools/codex/src/codex/searcher.py
"""Vector search for semantic code search."""

import json
import math
from typing import List, Optional

from .indexer import CodeIndex


def cosine_similarity(a: List[float], b: List[float]) -> float:
    """Calculate cosine similarity between two vectors."""
    if not a or not b:
        return 0.0

    # Handle different vector sizes (TF-IDF vectors may vary)
    min_len = min(len(a), len(b))
    a = a[:min_len]
    b = b[:min_len]

    dot_product = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))

    if norm_a == 0 or norm_b == 0:
        return 0.0

    return dot_product / (norm_a * norm_b)


class SearchResult:
    """A search result with ranking information."""

    def __init__(
        self,
        file_path: str,
        start_line: int,
        end_line: int,
        content: str,
        chunk_type: str,
        name: str,
        similarity: float,
    ):
        self.file_path = file_path
        self.start_line = start_line
        self.end_line = end_line
        self.content = content
        self.chunk_type = chunk_type
        self.name = name
        self.similarity = similarity

    @property
    def location(self) -> str:
        """Get file:line location string."""
        return f"{self.file_path}:{self.start_line}"

    @property
    def preview(self) -> str:
        """Get a preview of the content."""
        # First line or first 100 chars
        first_line = self.content.split("\n")[0].strip()
        if len(first_line) > 100:
            return first_line[:97] + "..."
        return first_line

    def __repr__(self) -> str:
        pct = self.similarity * 100
        return f"SearchResult({pct:.0f}% {self.location} {self.name})"


class CodeSearcher:
    """Search the code index."""

    def __init__(self, index: CodeIndex):
        self.index = index
        self.embedder = index.embedder

    def search(
        self,
        query: str,
        limit: int = 10,
        file_filter: Optional[str] = None,
        type_filter: Optional[str] = None,
        min_similarity: float = 0.1,
    ) -> List[SearchResult]:
        """
        Search for code matching the query.

        Args:
            query: Natural language query
            limit: Maximum number of results
            file_filter: Only search in files matching this pattern
            type_filter: Only search chunks of this type (function, class, etc.)
            min_similarity: Minimum similarity threshold

        Returns:
            List of SearchResult objects sorted by similarity
        """
        # Embed the query
        query_embedding = self.embedder.embed(query)

        if not query_embedding:
            return []

        # Build SQL query
        sql = "SELECT file_path, start_line, end_line, content, chunk_type, name, embedding FROM chunks WHERE 1=1"
        params = []

        if file_filter:
            sql += " AND file_path LIKE ?"
            params.append(f"%{file_filter}%")

        if type_filter:
            sql += " AND chunk_type = ?"
            params.append(type_filter)

        cursor = self.index.conn.execute(sql, params)

        # Calculate similarities
        results = []
        for row in cursor:
            try:
                chunk_embedding = json.loads(row["embedding"])
            except (json.JSONDecodeError, TypeError):
                continue

            similarity = cosine_similarity(query_embedding, chunk_embedding)

            if similarity >= min_similarity:
                results.append(SearchResult(
                    file_path=row["file_path"],
                    start_line=row["start_line"],
                    end_line=row["end_line"],
                    content=row["content"],
                    chunk_type=row["chunk_type"],
                    name=row["name"],
                    similarity=similarity,
                ))

        # Sort by similarity (descending)
        results.sort(key=lambda r: r.similarity, reverse=True)

        return results[:limit]

    def search_file(self, query: str, file_path: str, limit: int = 5) -> List[SearchResult]:
        """Search within a specific file."""
        return self.search(query, limit=limit, file_filter=file_path)

    def find_similar(
        self,
        file_path: str,
        line: int,
        limit: int = 5,
    ) -> List[SearchResult]:
        """
        Find code similar to a specific location.

        Args:
            file_path: Path to the reference file
            line: Line number in the reference file
            limit: Maximum number of results

        Returns:
            List of similar code chunks
        """
        # Find the chunk containing this line
        cursor = self.index.conn.execute("""
            SELECT content, embedding FROM chunks
            WHERE file_path = ? AND start_line <= ? AND end_line >= ?
            LIMIT 1
        """, (file_path, line, line))

        row = cursor.fetchone()
        if not row:
            return []

        try:
            reference_embedding = json.loads(row["embedding"])
        except (json.JSONDecodeError, TypeError):
            return []

        # Search for similar chunks
        cursor = self.index.conn.execute("""
            SELECT file_path, start_line, end_line, content, chunk_type, name, embedding
            FROM chunks
            WHERE NOT (file_path = ? AND start_line <= ? AND end_line >= ?)
        """, (file_path, line, line))

        results = []
        for row in cursor:
            try:
                chunk_embedding = json.loads(row["embedding"])
            except (json.JSONDecodeError, TypeError):
                continue

            similarity = cosine_similarity(reference_embedding, chunk_embedding)

            if similarity > 0.1:
                results.append(SearchResult(
                    file_path=row["file_path"],
                    start_line=row["start_line"],
                    end_line=row["end_line"],
                    content=row["content"],
                    chunk_type=row["chunk_type"],
                    name=row["name"],
                    similarity=similarity,
                ))

        results.sort(key=lambda r: r.similarity, reverse=True)
        return results[:limit]


def format_results(results: List[SearchResult], show_content: bool = False) -> str:
    """Format search results for display."""
    if not results:
        return "No results found."

    lines = []
    for i, r in enumerate(results, 1):
        pct = r.similarity * 100
        lines.append(f"{i}. [{pct:.0f}%] {r.location}")
        lines.append(f"   {r.chunk_type}: {r.name}")

        if show_content:
            # Show first few lines
            content_lines = r.content.split("\n")[:5]
            for line in content_lines:
                lines.append(f"   | {line[:80]}")
            if len(r.content.split("\n")) > 5:
                lines.append("   | ...")

        lines.append("")

    return "\n".join(lines)

File: crates/codex/src/searcher.rs
//! Search functionality for the code index.
//!
//! Uses SQLite FTS5 for keyword-based search as a fallback when
//! semantic embeddings are not available.

use anyhow::Result;
use rusqlite::{params, Connection};

use crate::indexer::CodeIndex;

/// A search result with ranking information
#[derive(Debug, Clone)]
pub struct SearchResult {
    /// Path relative to project root
    pub file_path: String,
    /// Starting line number
    pub start_line: usize,
    /// Ending line number
    pub end_line: usize,
    /// The code content
    pub content: String,
    /// Type of chunk
    pub chunk_type: String,
    /// Name of the function/class/etc
    pub name: String,
    /// Relevance score (higher is better)
    pub score: f64,
}

impl SearchResult {
    /// Get file:line location string
    pub fn location(&self) -> String {
        format!("{}:{}", self.file_path, self.start_line)
    }

    /// Get a preview of the content (first line, truncated)
    pub fn preview(&self) -> String {
        let first_line = self.content.lines().next().unwrap_or("").trim();
        if first_line.len() > 100 {
            format!("{}...", &first_line[..97])
        } else {
            first_line.to_string()
        }
    }
}

/// Search the code index
pub struct CodeSearcher<'a> {
    index: &'a mut CodeIndex,
}

impl<'a> CodeSearcher<'a> {
    /// Create a new searcher for the given index
    pub fn new(index: &'a mut CodeIndex) -> Self {
        Self { index }
    }

    /// Search for code matching the query
    pub fn search(
        &mut self,
        query: &str,
        limit: usize,
        file_filter: Option<&str>,
        type_filter: Option<&str>,
    ) -> Result<Vec<SearchResult>> {
        let conn = Connection::open(&self.index.db_path)?;

        // Build the FTS5 query
        // Escape special characters and add wildcards for partial matching
        let fts_query = self.build_fts_query(query);

        let mut sql = String::from(
            r#"
            SELECT
                c.file_path,
                c.start_line,
                c.end_line,
                c.content,
                c.chunk_type,
                c.name,
                bm25(chunks_fts) as score
            FROM chunks_fts
            JOIN chunks c ON chunks_fts.rowid = c.id
            WHERE chunks_fts MATCH ?
            "#,
        );

        let mut params_vec: Vec<Box<dyn rusqlite::ToSql>> = vec![Box::new(fts_query.clone())];

        if let Some(filter) = file_filter {
            sql.push_str(" AND c.file_path LIKE ?");
            params_vec.push(Box::new(format!("%{}%", filter)));
        }

        if let Some(filter) = type_filter {
            sql.push_str(" AND c.chunk_type = ?");
            params_vec.push(Box::new(filter.to_string()));
        }

        sql.push_str(" ORDER BY score LIMIT ?");
        params_vec.push(Box::new(limit as i64));

        let params_refs: Vec<&dyn rusqlite::ToSql> = params_vec.iter().map(|p| p.as_ref()).collect();

        let mut stmt = conn.prepare(&sql)?;
        let results = stmt
            .query_map(params_refs.as_slice(), |row| {
                Ok(SearchResult {
                    file_path: row.get(0)?,
                    start_line: row.get::<_, i64>(1)? as usize,
                    end_line: row.get::<_, i64>(2)? as usize,
                    content: row.get(3)?,
                    chunk_type: row.get(4)?,
                    name: row.get(5)?,
                    score: row.get::<_, f64>(6)?.abs(), // BM25 returns negative scores
                })
            })?
            .filter_map(|r| r.ok())
            .collect();

        Ok(results)
    }

    /// Build an FTS5 query from a natural language query
    fn build_fts_query(&self, query: &str) -> String {
        // Split query into words and create an OR query with prefix matching
        let words: Vec<&str> = query
            .split(|c: char| !c.is_alphanumeric() && c != '_')
            .filter(|w| w.len() > 1)
            .collect();

        if words.is_empty() {
            return query.to_string();
        }

        // Create prefix match for each word
        let terms: Vec<String> = words
            .iter()
            .map(|w| format!("\"{}\"*", w))
            .collect();

        // Use OR to match any of the terms
        terms.join(" OR ")
    }

    /// Search within a specific file
    pub fn search_file(&mut self, query: &str, file_path: &str, limit: usize) -> Result<Vec<SearchResult>> {
        self.search(query, limit, Some(file_path), None)
    }

    /// Find code similar to a specific location
    pub fn find_similar(
        &mut self,
        file_path: &str,
        line: usize,
        limit: usize,
    ) -> Result<Vec<SearchResult>> {
        let conn = Connection::open(&self.index.db_path)?;

        // Find the chunk at the specified location
        let reference_content: Option<String> = conn
            .query_row(
                r#"
                SELECT content FROM chunks
                WHERE file_path = ? AND start_line <= ? AND end_line >= ?
                LIMIT 1
                "#,
                params![file_path, line as i64, line as i64],
                |row| row.get(0),
            )
            .ok();

        let reference = match reference_content {
            Some(c) => c,
            None => return Ok(Vec::new()),
        };

        // Extract keywords from the reference content
        let keywords: Vec<&str> = reference
            .split(|c: char| !c.is_alphanumeric() && c != '_')
            .filter(|w| w.len() > 2)
            .take(20)
            .collect();

        if keywords.is_empty() {
            return Ok(Vec::new());
        }

        // Build FTS query from keywords
        let fts_query: String = keywords
            .iter()
            .map(|w| format!("\"{}\"", w))
            .collect::<Vec<_>>()
            .join(" OR ");

        // Search for similar content, excluding the reference chunk
        let mut stmt = conn.prepare(
            r#"
            SELECT
                c.file_path,
                c.start_line,
                c.end_line,
                c.content,
                c.chunk_type,
                c.name,
                bm25(chunks_fts) as score
            FROM chunks_fts
            JOIN chunks c ON chunks_fts.rowid = c.id
            WHERE chunks_fts MATCH ?
            AND NOT (c.file_path = ? AND c.start_line <= ? AND c.end_line >= ?)
            ORDER BY score
            LIMIT ?
            "#,
        )?;

        let results = stmt
            .query_map(
                params![fts_query, file_path, line as i64, line as i64, limit as i64],
                |row| {
                    Ok(SearchResult {
                        file_path: row.get(0)?,
                        start_line: row.get::<_, i64>(1)? as usize,
                        end_line: row.get::<_, i64>(2)? as usize,
                        content: row.get(3)?,
                        chunk_type: row.get(4)?,
                        name: row.get(5)?,
                        score: row.get::<_, f64>(6)?.abs(),
                    })
                },
            )?
            .filter_map(|r| r.ok())
            .collect();

        Ok(results)
    }
}

/// Format search results for display
pub fn format_results(results: &[SearchResult], show_content: bool) -> String {
    if results.is_empty() {
        return "No results found.".to_string();
    }

    let mut lines = Vec::new();

    for (i, result) in results.iter().enumerate() {
        let score_pct = (result.score * 100.0).min(100.0);
        lines.push(format!("{}. [{:.0}] {}", i + 1, score_pct, result.location()));
        lines.push(format!("   {}: {}", result.chunk_type, result.name));

        if show_content {
            for line in result.content.lines().take(5) {
                let truncated = if line.len() > 80 {
                    format!("{}...", &line[..77])
                } else {
                    line.to_string()
                };
                lines.push(format!("   | {}", truncated));
            }
            if result.content.lines().count() > 5 {
                lines.push("   | ...".to_string());
            }
        }

        lines.push(String::new());
    }

    lines.join("\n")
}

/// Format search results as JSON
pub fn format_results_json(results: &[SearchResult]) -> Result<String> {
    use serde_json::json;

    let items: Vec<_> = results
        .iter()
        .map(|r| {
            json!({
                "file_path": r.file_path,
                "start_line": r.start_line,
                "end_line": r.end_line,
                "chunk_type": r.chunk_type,
                "name": r.name,
                "score": r.score,
                "content": r.content,
            })
        })
        .collect();

    Ok(serde_json::to_string_pretty(&json!({
        "results": items,
        "count": results.len(),
    }))?)
}
